File: C/mpiio_master_vector.c
#include <stdio.h>
#include <stdlib.h>

#include "mpi.h"

#include "ioutils.h"

#include "arralloc.h"

#define NDIM 2

int main(int argc, char **argv){
  /*
   *  pcoords stores the grid positions of each process
   */

  int **pcoords;
  
  /*
   *  buf is the large buffer for the master to read into
   *  x contains the local data only
   */
  float **buf;
  float **x;

  int rank, size, dest, tag;
  int i, j;

  int nx, ny, nxp, nyp, xprocs, yprocs, barrier;

  MPI_Status status;

  int count, blocklength, stride;

  MPI_Datatype my_mpi_vector;

  int istart, jstart;

  double starttime, endtime, totaltime;

  char filename[MAXFILENAME];

  MPI_Comm comm = MPI_COMM_WORLD;

  MPI_Init(&argc, &argv);

  MPI_Comm_size(comm, &size);
  MPI_Comm_rank(comm, &rank);

  checkandgetarguments(argc, argv, &nx, &ny, &xprocs, &yprocs, &nxp, &nyp, &barrier, size, rank);

  pcoords = (int **)arralloc(sizeof(int), 2, size, NDIM);
  if(rank == 0){
    buf = (float **)arralloc(sizeof(float),2,nx,ny);
  }
  x = (float **)arralloc(sizeof(float),2,nxp,nyp);

  initpgrid(&pcoords[0][0], xprocs, yprocs);

  if (rank == 0)
    {
      printf("Running on %d process(es) in a %d x %d grid\n",
	     size, xprocs, yprocs);
      printf("\n");

      for (i=0; i < size; i++)
	{
	  printf("Process %2d has grid coordinates (%2d, %2d)\n",
		 i, pcoords[i][0], pcoords[i][1]);
	}
      printf("\n");
    }

  /*
   *  Initialise the arrays to a grey value
   */

  if(rank == 0){
    initarray(&buf[0][0], nx,  ny );
  }
  initarray(&x[0][0]  , nxp, nyp);

  if(barrier){
    MPI_Barrier(comm);
  }

  starttime = MPI_Wtime();

  /*
   *  Read the entire array on the master process
   *  Passing "-1" as the rank argument means that the file name has no
   *  trailing "_rank" appended to it, ie we read the global file
   */

  if (rank == 0)
    {
      ioread (argv[1], &buf[0][0], nx*ny);
      printf("\n");
    }

  tag = 0;

  /*
   *  Define the appropriate derived vector type
   *  The pattern is the same on all processes so everyone can make
   *  the same call
   */

  count = nxp;
  blocklength = nyp;
  stride = ny;

  MPI_Type_vector(count, blocklength, stride, MPI_FLOAT, &my_mpi_vector);

  /*  Commit it before use */

  MPI_Type_commit(&my_mpi_vector);


  if (rank == 0)
    {

      /*
       *  Send data to each of the workers in turn using synchronous sends
       *  Can do the sends in the usual order as we avoid the copy (except
       *  for the master)
       */

      for (dest = 1; dest < size; dest++)
	{

	  /*  Need to work out where to place the vector type within buf  */

	  istart = pcoords[dest][0]*nxp;
	  jstart = pcoords[dest][1]*nyp;

	  printf("rank %d sending to rank %d\n", rank, dest);

	  /*  Note that we send one vector */

	  MPI_Ssend(&(buf[istart][jstart]), 1, my_mpi_vector, dest, tag, comm);
	}

      /*  Master still needs to do a copy himself  */

      for (i=0; i < nxp; i++)
	{
	  for (j=0; j < nyp; j++)
	    {
	      x[i][j] = buf[i][j];
	    }
	}
    }
  else
    {
      /*
       *  Workers receive data from master
       */

      printf("rank %d receiving from rank 0\n", rank);

      /*  Note that we receive nxp*nyp reals */

      MPI_Recv(&x[0][0], nxp*nyp, MPI_FLOAT, 0, tag, comm, &status);
    }

  endtime = MPI_Wtime();

  /*
   *  Every process writes out its local data array x to an individually
   *  named file which has the rank appended to the file name
   */

#ifdef DEBUG
 createfilename(filename, "coutput", nxp, nyp, rank);
  iowrite(filename, &x[0][0], nxp*nyp);
#endif

  totaltime = endtime - starttime;

  dotimings(totaltime, rank, size);

  MPI_Type_free(&my_mpi_vector);

  MPI_Finalize();

  free(x);
  if(rank == 0){
    free(buf);
  }
  free(pcoords);
}

File: F/mpiio_master_vector.f90
program mpiio

  use mpi

  implicit none

!
!  The global data size is nx x ny
!

  integer :: nx
  integer :: ny

!
!  The processes are in a 2D array of dimension XPROCS x YPROCS, with
!  a total of NPROCS processes
!

  integer, parameter :: ndim = 2

  integer :: xprocs
  integer :: yprocs

!
!  The local data size is NXP x NYP
!

  integer :: nxp
  integer :: nyp

!
!  The maximum length of a file name
!

  integer, parameter :: maxfilename = 64

!
!  pcoords stores the grid positions of each process
!

  integer, dimension(:,:), allocatable :: pcoords

!
!  buf is the large buffer for the master to read into
!  x contains the local data only
!

  real, dimension(:,:), allocatable :: buf
  real, dimension(:,:), allocatable :: x

  integer :: rank, size, ierr, dest, tag, barrier
  integer :: i, j

  integer, dimension(MPI_STATUS_SIZE) :: status
  integer :: count, blocklength, stride
  integer :: my_mpi_vector

  integer :: istart, jstart

  character*(maxfilename) :: filename

  double precision :: starttime, endtime, totaltime

  integer :: comm = MPI_COMM_WORLD

  call MPI_INIT(ierr)

  call MPI_COMM_SIZE(comm, size, ierr)
  call MPI_COMM_RANK(comm, rank, ierr)

  call checkandgetarguments(filename, maxfilename, nx, ny, xprocs, yprocs, nxp, nyp, barrier, size, rank)

  allocate(pcoords(ndim,size))
  if(rank .eq. 0) then
   allocate(buf(nx,ny))
  end if
  allocate(x(nxp,nyp))

!
!  Work out the coordinates of all the processes in the grid and
!  print them out
!

  call initpgrid(pcoords, xprocs, yprocs)

  if (rank .eq. 0) then

     write(*,*) 'Running on ', size, ' process(es) in a ', &
                xprocs, ' x ', yprocs, ' grid'
     write(*,*)

     do i = 0, size-1
       write(*,*) 'Process ', i, ' has grid coordinates (', &
                  pcoords(1,i+1), ', ', pcoords(2,i+1), ')'
     end do

     write(*,*)

  end if
  
!
!  Initialise the arrays to a grey value
!
  if(rank .eq. 0) then
    call initarray(buf, nx,  ny )
  end if
  call initarray(x,   nxp, nyp)

  if(barrier .eq. 1) then
     call MPI_Barrier(comm, ierr)
  end if

  starttime = mpi_wtime()

!
!  Read the entire array on the master process
!  Passing "-1" as the rank argument means that the file name has no
!  trailing "_rank" appended to it, ie we read the global file
!

  if (rank .eq. 0) then

    call ioread (filename, buf, nx*ny)
    write(*,*)

  end if

  tag = 0

!  Define the appropriate derived vector type
!  The pattern is the same on all processes so everyone can make
!  the same call

  count = nyp
  blocklength = nxp
  stride = nx

  call MPI_TYPE_VECTOR(count, blocklength, stride,    &
                       MPI_REAL, my_mpi_vector, ierr)

!  Commit it before use

  call MPI_TYPE_COMMIT(my_mpi_vector, ierr)


  if (rank == 0) then 

!  Send data to each of the workers in turn using synchronous sends
!  Can do the sends in the usual order as we avoid the copy (except
!  for the master)

     do dest = 1, size-1

!  Need to work out where to place the vector type within buf

        istart = pcoords(1, dest+1)*nxp
        jstart = pcoords(2, dest+1)*nyp


        write(*,*) 'rank ', rank, ' sending to rank ', dest

!  Note that we send one vector

        call MPI_SSEND(buf(istart+1, jstart+1), 1, my_mpi_vector, &
                              dest, tag, comm, ierr)

     end do

!  Master still needs to do a copy to himself
     do j = 1, nyp
        do i = 1, nxp

           x(i,j) = buf(i,j)

        end do
     end do

  else

!  Workers receive data from master

     write(*,*) 'rank ', rank, ' receiving from rank 0'

!  Note that we receive nxp*nyp reals

     call MPI_RECV(x, nxp*nyp, MPI_REAL, 0, tag, comm, status, ierr)

  end if

  endtime = mpi_wtime()
!
!  Every process writes out its local data array x to an individually
!  named file which as the rank appended to the file name
!

!  call createfilename(filename, 'foutput', nxp, nyp, rank)
!  call iowrite(filename, x, nxp*nyp)

  totaltime = endtime-starttime
  call dotimings(totaltime, rank, size)

  call MPI_FINALIZE(ierr)

  deallocate(pcoords,x)
  if(rank .eq. 0) then
    deallocate(buf)
  end if

end

